File: ThreemaShareExtension/ProgressViewController.m
//  _____ _
// |_   _| |_  _ _ ___ ___ _ __  __ _
//   | | | ' \| '_/ -_) -_) '  \/ _` |_
//   |_| |_||_|_| \___\___|_|_|_\__,_(_)
//
// Threema iOS Client
// Copyright (c) 2015-2021 Threema GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/>.

#import "ProgressViewController.h"
#import "RectUtil.h"
#import "BundleUtil.h"

@interface ProgressViewController ()

@property (nonatomic) CGFloat progress;
@property NSMutableDictionary *itemsToSend;
@property UIVisualEffectView *blurView;

@end

@implementation ProgressViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    _itemsToSend = [NSMutableDictionary dictionary];
    [_cancelButton setTitle: [BundleUtil localizedStringForKey:@"cancel"] forState:UIControlStateNormal];
}

- (void)setupColors {

    switch ([Colors getTheme]) {
        case ColorThemeDark:
        case ColorThemeDarkWork:
            self.view.tintColor = [Colors main];
            
            _contentView.backgroundColor = [Colors background];
            _label.textColor = [Colors fontNormal];
            
            [self darkenVisualEffectsView];
            break;
        case ColorThemeLight:
        case ColorThemeLightWork:
        case ColorThemeUndefined:
            break;
    }
    _contentView.layer.cornerRadius = 15.0;
}

- (void)darkenVisualEffectsView {
    CGRect rect = self.view.bounds;
    
    UIBlurEffectStyle blurStyle = UIBlurEffectStyleExtraLight;
    blurStyle = UIBlurEffectStyleDark;

    UIBlurEffect *blurEffect = [UIBlurEffect effectWithStyle:blurStyle];
    _blurView = [[UIVisualEffectView alloc] initWithEffect:blurEffect];
    
    _blurView.frame = rect;
    _blurView.autoresizingMask = UIViewAutoresizingFlexibleHeight | UIViewAutoresizingFlexibleWidth;
    
    [_visualEffectsView removeFromSuperview];
    _visualEffectsView = nil;

    [_blurView.contentView addSubview:_contentView];
    [self.view addSubview:_blurView];

    NSLayoutConstraint *horizontal = [NSLayoutConstraint constraintWithItem:_contentView
                                                                  attribute:NSLayoutAttributeCenterX
                                                                  relatedBy:NSLayoutRelationEqual
                                                                     toItem:_blurView
                                                                  attribute:NSLayoutAttributeCenterX
                                                                 multiplier:1.f constant:0.f];
    
    NSLayoutConstraint *vertical = [NSLayoutConstraint constraintWithItem:_contentView
                                                                attribute:NSLayoutAttributeCenterY
                                                                relatedBy:NSLayoutRelationEqual
                                                                   toItem:_blurView
                                                                attribute:NSLayoutAttributeCenterY
                                                               multiplier:1.f constant:0.f];
    
    [_blurView addConstraints:@[horizontal, vertical]];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    self.navigationController.navigationBarHidden = YES;
    
    [self updateProgressLabel];
    _progressView.progress = 0.0;
    
    [self setupColors];
}

- (void)updateProgressLabel {
    NSInteger inProgressOrSentCount = 0;
    for (NSString *key in _itemsToSend.keyEnumerator) {
        NSNumber *progress = [_itemsToSend objectForKey:key];
        if (progress.floatValue > 0.0) {
            inProgressOrSentCount++;
        };
    }

    NSString *sendingText = NSLocalizedString(@"sending_count", nil);
    NSInteger currentItemCount = inProgressOrSentCount <= _totalCount ? inProgressOrSentCount : _totalCount;
    NSString *text = [NSString stringWithFormat:sendingText, currentItemCount, (long)_totalCount];
    [_label setText: text];
}

- (void)setProgress:(NSNumber *)progress forItem:(id)item {
    if (item) {
        [_itemsToSend setObject:progress forKey:item];
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        [self updateProgressLabel];
        [self updateProgressView];
    });
}

- (void)finishedItem:(id)item {
    [self setProgress:[NSNumber numberWithFloat:1.0] forItem:item];
}

- (void)updateProgressView {
    CGFloat progress = 0.0;
    for (NSString *key in _itemsToSend.keyEnumerator) {
        NSNumber *itemProgress = [_itemsToSend objectForKey:key];
        progress += itemProgress.floatValue;
    }
    
    _progressView.progress = progress / _totalCount;
}

- (IBAction)didCancel:(id)sender {
    [_delegate progressViewDidCancel];
}

@end

File: ThreemaShareExtension/ProgressViewController.swift
import CocoaLumberjackSwift
import ThreemaFramework
import ThreemaMacros
import UIKit

protocol ProgressViewDelegate {
    func progressViewDidCancel()
}

class ProgressViewController: UIViewController {
    
    var itemsToSend: NSMutableDictionary?
    var blurView: UIVisualEffectView?
    
    var totalCount = 0
    
    var delegate: ProgressViewDelegate?
    
    @IBOutlet var label: UILabel!
    @IBOutlet var contentView: UIView!
    @IBOutlet var visualEffectsView: UIVisualEffectView!
    @IBOutlet var cancelButton: UIButton!
    @IBOutlet var progressView: UIProgressView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        itemsToSend = NSMutableDictionary()
        cancelButton!.setTitle(#localize("cancel"), for: .normal)
                
        view.backgroundColor = .systemGroupedBackground
        
        view.tintColor = .tintColor
        contentView!.backgroundColor = .secondarySystemGroupedBackground
        contentView?.layer.cornerRadius = 15.0
        label.textColor = .label
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        navigationController?.isNavigationBarHidden = true
        
        updateProgressLabel()
        progressView?.progress = 0.0
    }
    
    func setProgress(progress: NSNumber, item: Any?) {
        if item != nil {
            itemsToSend?.setObject(progress, forKey: item as! NSCopying)
        }
        
        DispatchQueue.main.async {
            self.updateProgressLabel()
            self.updateProgressView()
        }
    }
    
    func finishedItem(item: Any) {
        setProgress(progress: NSNumber(floatLiteral: 1.0), item: item)
    }
    
    private func updateProgressLabel() {
        var inProgressOrSentCount = 0
        var sentCount = 0
        for key in itemsToSend!.keyEnumerator() {
            let progress: NSNumber = itemsToSend!.object(forKey: key) as! NSNumber
            if progress.floatValue > 0.0 {
                inProgressOrSentCount = inProgressOrSentCount + 1
            }
            if progress.floatValue == 1.0 {
                sentCount += 1
            }
        }
        
        let currentItemCount = inProgressOrSentCount <= totalCount ? inProgressOrSentCount : totalCount
        
        var text = ""
        
        if sentCount == totalCount {
            text = #localize("finished_sending_title")
        }
        else {
            let sendingText = #localize("sending_count")
            text = String.localizedStringWithFormat(sendingText, currentItemCount, totalCount)
        }
        
        label!.text = text
    }
    
    private func updateProgressView() {
        var progress: Float = 0.0
        for key in itemsToSend!.keyEnumerator() {
            let itemProgress: NSNumber = itemsToSend?.object(forKey: key) as! NSNumber
            progress += itemProgress.floatValue
        }
        
        progressView?.progress = progress / Float(totalCount)
    }
    
    @IBAction func cancelButtonPressed(_ sender: Any) {
        guard let rootNavController = delegate else {
            let message = "Delegate was unexpecedly nil"
            DDLogError("\(message)")
            fatalError(message)
        }
        rootNavController.progressViewDidCancel()
    }
}
